common/src/main/java/tachyon/collections/IndexedSet.java

```diff
@@ -194,14 +208,54 @@ public boolean add(T object) {
   /**
    * Returns an iterator over the elements in this set. The elements are returned in no particular
    * order. It is to implement {@link Iterable} so that users can foreach the IndexedSet directly.
-   * The traversal may reflect any modifications subsequent to the construction of the iterator.
+   *
+   * Note that the behaviour of the iterator is unspecified if the underlying collection is
+   * modified while a thread is going through the iterator.
    *
    * @return an iterator over the elements in this IndexedSet
    */
   @Override
   public Iterator<T> iterator() {
-    synchronized (mLock) {
-      return mObjects.iterator();
+    return new IndexedSetIterator();
+  }
+
+  /**
+   *  Specialized iterator for {@link IndexedSet}.
+   *
+   *  This is needed to support consistent removal from the set and the indices.
+   */
+  private class IndexedSetIterator implements Iterator<T> {
+    private final Iterator<T> mSetIterator;
+    private T mLast;
+
+    public IndexedSetIterator() {
+      mSetIterator = mObjects.iterator();
+      mLast = null;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return mSetIterator.hasNext();
+    }
+
+    @Override
+    public T next() {
+      final T next = mSetIterator.next();
+      mLast = next;
+      return next;
+    }
+
+    @Override
+    public void remove() {
+      if (mLast != null) {
```

Comment:
NIT: maybe just do `Preconditions.checkState(mLast != null, "Next was never called before")` in the beginning of this method.